### PR TITLE
Add Error Serializers for custom and nested errors

### PR DESCRIPTION
**Pull request by n8sabes**
# What is it?

- [X] Feature / enhancement
- [X] Bug
- [ ] Docs / tests

# Description

<h1>This PR enables the serialization and deserialization of Error and Custom Error objects.
It serializes and deserializes primitive values and rebuilds nested Error objects in a targeted fashion.</h1>

The `serializeError` function takes an Error or Custom Error object as input and clones its primitive values using the `cloneErrorPrimitives` function before returning the resulting object in a JSON string format.

The `deserializeError` function takes a serialized Error object string as input and converts it back into an Error or Custom Error object with only the necessary primitive values. It accomplishes this by parsing the input string into an object, extracting the needed values and creating a new Error object with those values. If the input string contains nested Error objects, it will rebuild them by calling the `deserializeDeepErrors` function.

The `cloneErrorPrimitives` function recursively extracts all primitive values from an object or array, including handling of circular references, pruning the stack trace from Error objects, and returning a new object with only primitive values. This function is called by `serializeError` to extract the necessary values from the Error object.

Finally, the `deserializeDeepErrors` function rebuilds any nested Error objects that were serialized. It takes the rebuilt Error objects and inserts them back into the appropriate locations within the parent Error object.

# Use cases and why

<h2>As-is implemented, a simple string is *NOT* enough.</h2>

1. Complete Error and Custom errors are necessary when working with many backend systems.
2. Some backend systems such as GraphQL servers even return nested errors arrays which would be lost with the existing serializer, reducing the top level error to a string that is recreated on deserialization.
3. Other systems use fields like `code` and `status` which are not supported.
4. Other workarounds like simply using `JSON.stringify()` and `JSON.parse()` on an error or custom error objects is not natively simple due to issues like: **a)** circular references, and **b)** needing to reconstitute the nested errors on deserialization.  

<strong>Related Conversation in Discord:</strong> [HERE](https://discord.com/channels/842438759945601056/1085107180384030750)

# Checklist:

- [X] My code follows the [developer guidelines of this project](https://github.com/BuilderIO/qwik/blob/main/CONTRIBUTING.md)
- [X] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [1/2 ] Added new tests to cover the fix / functionality

Validation object provided for custom error testing.

### CODE REVIEW

**Comment by n8sabes** [comment]
<h3>Test File</h3>
Here is an `index.tsx` test file that exercises the serializer. I have also used it in a staged system through `server$` boundaries.

```typescript
import { $, component$ } from '@builder.io/qwik';
import { deserializeError, serializeError } from '~/utils/errorSerializer';


class CustomError extends Error {
  custom: any;

  constructor(message: string, custom?: any) {
    super(message);
    this.custom = custom;
  }
}

export default component$(() => {

  const simpleErrorTest = $(() => {
    const xErr = new CustomError("Deep X-Error!", 1234);
    const serErr = serializeError(xErr);
    console.log("serErr:", serErr);
    const desErr = deserializeError(serErr);
    console.log("desErr:", desErr);
    console.log("desErr:", JSON.stringify(desErr));
  });

  const deepErrorTest = $(() => {
    const validate = {
      e1: new Error("Standard Error"),
      c: {},
      o: {
        e2: new CustomError("Custom Error", {
          e3: new Error("Super Deep Standard Error"),
        }),
        c: {},
        stack: "keep me",
        n: null,
        b: true,
        v: 123,
        s: "abc",
        u: undefined,
        f: () => "function!",
      },
      stack: "keep me",
      n: null,
      b: true,
      v: 123,
      s: "abc",
      u: undefined,
      f: () => "function!"
    };
    validate.c = validate;
    validate.o.c = validate;

    const xErr = new CustomError("Deep Error!", validate);
    const serErr = serializeError(xErr);
    console.log("serErr:", serErr);
    const desErr = deserializeError(serErr);
    console.log("desErr:", desErr);
    console.log("desErr:", JSON.stringify(desErr));
  });

  return (<div>
    <h1>Test Error Serialization</h1>
    <p>Serialize errors to JSON and back again. Stack and functions are not serialized and prevents circular reference.</p><br />
    <button onClick$={simpleErrorTest}>Simple Error Test</button><br /><br />
    <button onClick$={deepErrorTest}>Deep Error Test</button>
  </div>);
})
```

**Comment by zanettin** [comment]
Hey Sebastian 👋 
CI complains atm because `pnpm fmt` needs to be exec 👼 thanks for this PR 🙏

**Comment by n8sabes** [comment]
Thanks @zanettin, didn't realize `pnpm fmt` was a thing. Updated the checkin.

I have Prettier installed, but I can't figure out why it formats some `.ts` files with two indents and others with four. I have vscode defaults set to Prettier, even. It drives me crazy.

**Comment by n8sabes** [comment]
@manucorporat, I have added **Dynamic Error** Class generation during deserialization so that the `constructor.name` comes through.

**Comment by manucorporat** [comment]
This is amazing work! i am a little concerned about the amount runtime added just to solve this issue, need to run it with the team

**Comment by n8sabes** [comment]
<h2>OPTIMIZED -- PERF NOW SAME AS OLD OPTIMIZER FOR NATIVE ERRORS!</h2>

@manucorporat, thanks for the comment. I've gone back and reviewed the new error serializer, and in several places heavily optimized it to perform as well as the old serializer. The added advantage of this is it now support custom errors and deep object errors. -- Please take a looks. There may be was to further optimize it, and welcome you throughs. Please see the attached screenshot of benchmarks from test page (above).
New Error Serializer Test Results

**Comment by n8sabes** [comment]
<h2>OPTIMIZED PERFORMANCE</h2>
<h3>EQUAL TO OLD OPTIMIZER WITH BETTER NATIVE AND CUSTOM ERROR SUPPORT</h3>

@manucorporat, I've gone back and reviewed the new error serializer and optimized it to perform as well as the old serializer *(maybe a bit better)*. The added advantage of this is it now supports custom errors and deep error objects (like from GraphQL errors). There may be ways to further optimize it and welcome you throughs.  Classnames were not possible without evaluating a class inline which would be unsafe but `prototype.name` work well now. 

Please see the attached screenshot of benchmarks run with the test page provided. The page also has tests that can be migrated to a unit test.

New Error Serializer Test Results

**Comment by n8sabes** [comment]
<h2> New Error Serializer Test on StackBlitz </h2>

[New Error Serializer Test on StackBlitz](https://stackblitz.com/edit/qwik-starter-vuzfrv?file=src/routes/index.tsx)

New Error Serializer Test Results

**Comment by n8sabes** [comment]
<h1>NEW: Deep Error Mitigation</h1>
I had an idea to further optimize error serialization. In version 2, the error serializer **always** mitigates what error data is send back. A developer may not realize a custom error could have additional objects and there is no point in transferring them if they are not to be consumed.

To transfer errors with custom fields or deep objects, such as in GraphQL, the developer must add the `__qDeep` flag to the error object. If this flag is present, only then will the error with custom fields/objects be serialized over the wire. The `__qDeep` flag is automatically be stripped out during the deserialization process.

I have updated the StackBlitz ( [HERE](https://stackblitz.com/edit/qwik-starter-vuzfrv) ) with both versions to explore, but think I like Version 2 better to avoid unnecessary data serialization.

File: packages/qwik/src/core/error/error_serializer.ts [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    let deepErrObj = obj;
+    pathKeys.forEach((key, i) => {
+      objKey = key;
+      parentObj = deepErrObj;
+      deepErrObj = deepErrObj[pathKeys[i]];
+    });
+
+    // Rebuild the object with Error instance
+    if (objKey) {
+      if (deepErrObj.__qName) {
```
Comment (tasawar-hussain) [inline packages/qwik/src/core/error/error_serializer.ts:92]:
```suggestion
    if (objKey && deepErrObj.__qName) {
```

File: packages/qwik/src/core/error/error_serializer.ts [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        const { message, __qName, __qIsError, stack, ...rest }: any = deepErrObj;
+        stack;
```
Comment (tasawar-hussain) [inline packages/qwik/src/core/error/error_serializer.ts:94]:
```suggestion
```

**Review by tasawar-hussain — CHANGES_REQUESTED** [review]

**Pull request closed without merging**